src/bp/core/services/bot-service.ts

```diff
@@ -210,9 +220,70 @@ export class BotService {
       requested_by
     }
 
-    await this.configProvider.mergeBotConfig(botId, { pipeline_status: { stage_request } })
-    // TODO call hook promotion request here
-    // keep reference to old config, if config has changed then call the onPromotionHook
+    const currentBot = await this.configProvider.mergeBotConfig(botId, { pipeline_status: { stage_request } })
+    await this._executeStageChangeHooks(currentBot)
+  }
+
+  private async _executeStageChangeHooks(initialBot: BotConfig) {
+    const alteredBot = _.cloneDeep(initialBot)
+    const users = await this.workspaceService.listUsers(['email', 'role'])
+    const pipeline = await this.workspaceService.getPipeline()
+    const api = await createForGlobalHooks()
+    const currentStage = <Stage>pipeline.find(s => s.id === initialBot.pipeline_status.current_stage.id)
+    const hookResult = {
+      actions: [currentStage.action]
+    }
+
+    await this.hookService.executeHook(new Hooks.OnStageChangeRequest(api, alteredBot, users, pipeline, hookResult))
+    if (_.isArray(hookResult.actions)) {
+      await Promise.map(hookResult.actions, action => {
+        if (action === 'promote_copy') {
+          this._promoteCopy(initialBot, alteredBot)
+        } else if (action === 'promote_move') {
+          this._promoteMove(alteredBot)
+        }
+      })
+    }
+    // stage has changed
+    if (initialBot.pipeline_status.current_stage.id !== alteredBot.pipeline_status.current_stage.id) {
+      await this.hookService.executeHook(new Hooks.AfterStageChanged(api, initialBot, alteredBot, users, pipeline))
+    }
+  }
+
+  private async _promoteMove(bot: BotConfig) {
+    bot.pipeline_status.current_stage = {
+      id: bot.pipeline_status.stage_request!.id,
+      promoted_by: bot.pipeline_status.stage_request!.requested_by,
+      promoted_on: new Date()
+    }
+    delete bot.pipeline_status.stage_request
+    return this.configProvider.setBotConfig(bot.id, bot)
+  }
+
+  private async _promoteCopy(initialBot: BotConfig, newBot: BotConfig) {
+    if ((await this.workspaceService.getBotRefs()).find(bId => bId == newBot.id)) {
+      const d = new Date()
+        .toDateString()
+        .split(' ')
+        .join('-')
+        .toLowerCase()
+      newBot.id = `${newBot.id}-copy-${d}-${Date.now()}`
```

Comment:
What do you think of adding the ID logic in the hook, and returning it in the hookResult? It would give more flexibility. Ex, if they want to promote copy mybot-dev to mybot-staging, to mybot only, it would be easier

Reply:
That's what's been done here. This is just in case the dev didn't modify the bot id. I add a default bot id because we don't want conflicts